kernel/mmanutils.c

```diff
@@ -350,114 +454,36 @@ int myst_mprotect(const void* addr, const size_t len, const int prot)
-            uint8_t* hi = (uint8_t*)addr + length;
-            uint8_t* plo = p->addr;
-            uint8_t* phi = (uint8_t*)p->addr + p->length;
-            uint8_t* maxlo = _max_ptr(lo, plo);
-            uint8_t* minhi = _min_ptr(hi, phi);
-
-            /* if there is an overlap */
-            if (maxlo < minhi)
-            {
-                size_t llength = maxlo - plo;
-                size_t rlength = phi - minhi;
-                size_t roffset = p->offset + (minhi - plo);
-
-                //     .........
-                // ..........
-                //
-                // .........
-                //     ..........
-
-                // left range:  [plo:llength]
-                // right range: [maxhi:rlength]
-
-                if (llength && rlength)
-                {
-                    // printf("case1: split\n");
-                    msync_mapping_t* rm;
-
-                    /* create the right mapping */
-                    if (!(rm = _new_msync_mapping(
-                              p->fd,
-                              roffset,   /* offset */
-                              minhi,     /* addr */
-                              rlength))) /* length */
-                    {
-                        ERAISE(-ENOMEM);
-                    }
-
-                    /* insert right mapping into list */
-                    rm->next = p->next;
-                    p->next = rm;
-
-                    /* update the left mapping length */
-                    p->length = llength;
-                    prev = rm;
-                }
-                else if (llength)
-                {
-                    // printf("case2: left\n");
+    if (!addr || !length)
+        ERAISE(-EINVAL);
 
-                    /* update the left mapping length */
-                    p->length = llength;
-                    prev = p;
-                }
-                else if (rlength)
-                {
-                    // printf("case3: right\n");
-                    p->offset = roffset;
-                    p->addr = minhi;
-                    p->length = rlength;
-                    prev = p;
-                }
-                else
-                {
-                    // printf("case4: remove\n");
-                    if (prev)
-                        prev->next = p->next;
-                    else
-                        _msync_mappings = p->next;
+    ECHECK(myst_round_up(length, PAGE_SIZE, &length));
+    ECHECK((index = _get_page_index(addr, length)));
 
-                    free(p);
-                }
-            }
-            else
-            {
-                prev = p;
-            }
+    myst_rspin_lock(&_mman.lock);
+    {
+        const size_t count = length / PAGE_SIZE;
 
-            p = next;
+        for (size_t i = index; i < index + count; i++)
+        {
+            myst_fdmapping_t* p = &v.fdmappings[index];
+
+            /* remove this fd-mapping */
+            p->used = 0;
```

Comment:
assert if it was not mapped?

Reply:
It is legal for it not to be mapped. This is called by ``munmap()`` and it removes any file mappings if there are any (and there usually will not be).